nntrainer/src/addition_layer.cpp

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: Apache-2.0-only
+/**
+ * Copyright (C) 2020 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file	addition_layer.cpp
+ * @date	30 July 2020
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug		No known bugs except for NYI items
+ * @brief	This is Addition Layer Class for Neural Network
+ *
+ */
+
+#include <addition_layer.h>
+#include <layer.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+int AdditionLayer::initialize(bool last) {
+  int status = ML_ERROR_NONE;
+  if (input_dim.getDataLen() == 1) {
+    ml_logw("Warning: the length of previous layer dimension is one");
+  }
+
+  this->last_layer = last;
+
+  /** input dimension indicates the dimension for all the inputs to follow */
+  output_dim = input_dim;
+
+  return status;
+}
+
+sharedConstTensor AdditionLayer::forwarding(sharedConstTensor in) {
+  hidden = Tensor(in->batch(), output_dim.channel(), output_dim.height(),
+                  output_dim.width());
+  hidden.setZero();
+
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    if (input_dim != in.get()[idx].getDim())
+      throw std::runtime_error("Error: addition layer requires same "
+                               "shape from all input layers");
+    hidden.add_i(in.get()[idx]);
+  }
+
+  return MAKE_SHARED_TENSOR(hidden);
+}
+
+sharedConstTensor AdditionLayer::backwarding(sharedConstTensor derivative,
+                                             int iteration) {
+  sharedTensor ret = std::shared_ptr<Tensor>(new Tensor[num_inputs],
+                                             std::default_delete<Tensor[]>());
+
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    Tensor &t = ret.get()[idx];
+    t = *derivative;
+  }
+
+  return ret;
```

Comment:
can we return `drivative` instead of making `ret`?

Reply:
No, we need to return an array of the size of `num_inputs`. Each of these derivatives will be sent back to the corresponding previous layer. One can optimize this to be just 1 derivative and let the caller send each copy. But cannot guarantee the proper working of this layer.

Reply:
Right.